File: Aoc.Csharp/Solutions/Day06.cs
namespace Aoc.Solutions.Day06;
public record State(Dictionary<long, long> Fish)
{
    public State Next()
    {
        var nextState =
            Fish
            .SelectMany(kvp => kvp.Key switch
            {
                var k when k == 0 => new List<(long, long)> { (6, kvp.Value), (8, kvp.Value) },
                var k => new List<(long, long)> { (k - 1, kvp.Value) }
            })
            .GroupBy(fishLifeState => fishLifeState.Item1)
            .Select(fishLifeStateGroup => (fishLifeStateGroup.Key, fishLifeStateGroup.Select(v => v.Item2).Sum()))
            .ToDictionary(g => g.Key, g => g.Item2);

        return new(nextState);
    }
}

public class Day06 : Day
{
    public static State Parse(string line) =>
        line
            .Split(',')
            .Select(long.Parse)
            .GroupBy(v => v)
            .ToDictionary(v => v.Key, v => v.LongCount())
            .Then(dict => new State(dict));

    public override string SolveA(string input)
    {
        var state = Parse(input);
        foreach (var i in Range(0, 80))
        {
            state = state.Next();
        }
        return state.Fish.Values.Sum().ToString();
    }

    public override string SolveB(string input)
    {
        var state = Parse(input);
        foreach (var i in Range(0, 256))
        {
            state = state.Next();
        }
        return state.Fish.Values.Sum().ToString();
    }

    public Day06()
    {
        Tests = new()
        {
            new("A", "3,4,3,1,2", "5934", SolveA)
        };
    }
}

File: Aoc.Fsharp/Solutions/Day06.fs
module Aoc.Solutions.Day06

open Aoc.Runner

let parse (input: string) =
    input.Split(',')
    |> Seq.map int64
    |> Seq.groupBy (fun x -> x)
    |> Seq.map (fun (key, values) -> (key, Seq.length values |> int64))

let nextState state =
    state
    |> Seq.map (fun (key, values) ->
        seq {
            if key = 0L then
                yield (6L, values)
                yield (8L, values)
            else
                yield (key - 1L, values)
        })
    |> Seq.concat
    |> Seq.groupBy (fun (key, _) -> key)
    |> Seq.map (fun (key, multValues) -> (key, multValues |> Seq.map (fun (k, v) -> v) |> Seq.sum))

type Day06() =
    inherit Day()

    override _.SolveA input =
        let state = parse input

        seq { 1L .. 80L }
        |> Seq.fold (fun cur _ -> nextState cur) state
        |> Seq.sumBy (fun (_, v) -> v)
        |> string

    override _.SolveB input =
        let state = parse input

        seq { 1L .. 256L }
        |> Seq.fold (fun cur _ -> nextState cur) state
        |> Seq.sumBy (fun (_, v) -> v)
        |> string
